students/TacSPx/lesson04/mailroompart2.py

```diff
@@ -0,0 +1,216 @@
+            else:
+                donor_info.setdefault(donor_name, []).append(donor_amount)
+                print(f"New donor name '{donor_name}' added to registry!")
+                input("\nPress enter to continue to 'Thank You' letter:")
+                # print(donor_info)  # - for testing
+                print(letter(donor_name, donor_amount))
+                break
+            print(letter(donor_name, donor_amount))
+            break
+
+def letter(donor_name, donor_amount):
+
+    new_letter = f'''Dear {donor_name},
+
+Thank you for your recent donation of '${donor_amount}' to XYZ Nonprofit for children! It really makes a huge impact for
+the children in our community.
+
+Those three hours after the end of the school day can make a crucial difference in kids’ lives.
+Thanks to you, kids have a safe place to go after school. Instead of going home alone while their families are at work,
+our kids are learning to play sports, create art, and improving their grades at our Homework Help Center.  All while
+forming friendships with peers and relationships with adult mentors and tutors.
+
+Thank you again {donor_name}, for your ongoing support of our kids!
+
+Sincerely,
+
+XYZ Nonprofit Agency Director
+'''
+    return new_letter
+###############################################
+#   #2 Create a Report
+###############################################
+def report():
+    ''' Report: creates a tabular summary report of donor transactions'''
+    while True:
+        donor_sort = {}
+        donor_sum = {}
+        donor_len = {}
+        donor_avg = {}
+        donor_merge = {}
+        # title
+        title = "*******************  XYZ Nonprofit Charity  ********************\n" \
+                "________________________________________________________________"
+        heading_name = ("| Donor Name:", "| Totals $:", "| #-of-Gifts:", "| Avg Gift $: ", "|")
+        header = f"{heading_name[0].upper():<20s} {heading_name[1].upper():<13s}" \
+                 f" {heading_name[2].upper():<6s} {heading_name[3].upper():<14s}|"
+        print(title)
+        print(header)
+        # define format row
+        row = "| {Name:<18s} | ${Total:<10.2f} | {NumbGifts:^11d} | ${AverageGifts:<10.2f} |".format
+        # revisions made below to add sorting by totals, step #1 set-up to sort, new list
+
+        # calcs in the dictionary
+        donor_sum = dict([(key, sum(values)) for key, values in donor_info.items()])
+        donor_len = dict([(key, len(values)) for key, values in donor_info.items()])
+        donor_avg = dict([(key, sum(values) / len(values)) for key, values in donor_info.items()])
+        # merge
+        for key in set(list(donor_sum.keys()) + list(donor_len.keys()) + list(donor_avg.keys())):
+            donor_merge.setdefault(key, []).append(donor_sum[key])
+            donor_merge.setdefault(key, []).append(donor_len[key])
+            donor_merge.setdefault(key, []).append(donor_avg[key])
```

Comment:
not sure on the purpose of the above 3 dictionaries and merging them - you shouldn't need any of that

pass donor_info to sorted() function and sort on the sum